git-traverse/src/commit.rs

```diff
@@ -109,6 +167,63 @@ pub mod ancestors {
             res.map(Ok)
         }
     }
+
+    mod predicates {
+        use super::*;
+
+        /// `FnMut`-like trait used for pruning commit subgraphs from the ancestors iterator.
+        ///
+        /// This trait only exists because:
+        /// 1. Predicate usage should be optional.
+        ///    The user should opt-in to using predicates via an alternate constructor or via the
+        ///    builder pattern.
+        /// 2. The default case (include all ancestors) should not increase the size of the
+        ///    [`Ancestors`] struct, nor should it require extra branches at runtime.
+        ///    So don't use boxed closures, `Option<impl FnMut...>`, or `fn(&oid) -> bool`.
+        ///    Instead, use a zero-sized type ([`AlwaysTrue`]) to handle the default case.
+        /// 3. Directly implementing `FnMut` is not yet supported in stable Rust.
+        /// 4. I could not get the compiler to accept [`Ancestors::new`] using a closure or static
+        ///    function as the default predicate. I don't think this is possible without using
+        ///    something like `typeof!(always_true)` (where `always_true` is a static function) as a
+        ///    type parameter.
+        pub trait OidPredicate {
```

Comment:
So from what I can tell, `OidPredicate` is effectively a reimplementation of `FnMut`?

I'm a little confused w.r.t the 3rd point: what limitation are we running into with trying to directly use `FnMut`? We do something similar with the `Find` type (e.g. a `for<'a> FnMut`) so I would expect something similar to work?

Reply:
Yes, `OidPredicate` is a reimplementation of `FnMut`.

I could not get a type parameter like `where Predicate: FnMut(&oid) -> bool` to work in an `impl Ancestors` block where `Ancestors::new` itself tried to supply the default value for `predicate`. The compiler complained that the default value supplied by `Ancestors::new` was not guaranteed to fit the caller-chosen type for `Predicate`.

From that error, I gather that the `impl Ancestors` block containing `Ancestors::new` needs to force a specific type for `Predicate`. I don't know of a way to get the exact type of a closure or static function to use in `impl<...> Ancestors<Find, ???, StateMut>` that doesn't rely on non-zero-cost trait objects or function pointers, so I went for a struct that could impl `FnMut` directly.

But then I hit the issue where trying to make a struct impl `FnMut` directly requires using `#![feature(fn_traits)]`, and I figured this was a non-starter. So `OidPredicate` was born.

`Find` should hit the same problem if you tried to provide a constructor that supplied a default value for it.

The closest discussion I found find related to this is https://stackoverflow.com/a/37313554.